src/nodes.coffee

```diff
@@ -380,59 +381,63 @@ exports.Block = class Block extends Base
 
 #### Literal
 
-# Literals are static values that can be passed through directly into
-# JavaScript without translation, such as: strings, numbers,
+# `Literal` is a base class for static values that can be passed through
+# directly into JavaScript without translation, such as: strings, numbers,
 # `true`, `false`, `null`...
 exports.Literal = class Literal extends Base
   constructor: (@value) ->
 
-  makeReturn: ->
-    if @isStatement() then this else super
-
-  isAssignable: ->
-    IDENTIFIER.test @value
-
-  isStatement: ->
-    @value in ['break', 'continue', 'debugger']
-
   isComplex: NO
 
   assigns: (name) ->
     name is @value
 
+  compileNode: (o) ->
+    [@makeCode @value]
+
+  toString: ->
+    " #{if @isStatement() then super else @constructor.name}: #{@value}"
+
+exports.NumberLiteral = class NumberLiteral extends Literal
+
+exports.StringLiteral = class StringLiteral extends Literal
+
+exports.RegexLiteral = class RegexLiteral extends Literal
+
+exports.PassthroughLiteral = class PassthroughLiteral extends Literal
+
+exports.IdentifierLiteral = class IdentifierLiteral extends Literal
+  isAssignable: -> @value not in RESERVED
+
+exports.StatementLiteral = class StatementLiteral extends Literal
+  isStatement: YES
+
+  makeReturn: THIS
+
   jumps: (o) ->
     return this if @value is 'break' and not (o?.loop or o?.block)
     return this if @value is 'continue' and not o?.loop
 
   compileNode: (o) ->
-    code = if @value is 'this'
-      if o.scope.method?.bound then o.scope.method.context else @value
-    else if @value.reserved
-      "\"#{@value}\""
-    else
-      @value
-    answer = if @isStatement() then "#{@tab}#{code};" else code
-    [@makeCode answer]
+    [@makeCode "#{@tab}#{@value};"]
 
-  toString: ->
-    ' "' + @value + '"'
+exports.ThisLiteral = class ThisLiteral extends Literal
+  constructor: ->
+    super 'this'
 
-class exports.Undefined extends Base
-  isAssignable: NO
-  isComplex: NO
+  compileNode: (o) ->
+    code = if o.scope.method?.bound then o.scope.method.context else @value
+    [@makeCode code]
+
+class exports.Undefined extends Literal
+  constructor: -> super 'undefined'
   compileNode: (o) ->
     [@makeCode if o.level >= LEVEL_ACCESS then '(void 0)' else 'void 0']
 
-class exports.Null extends Base
-  isAssignable: NO
-  isComplex: NO
-  compileNode: -> [@makeCode "null"]
+class exports.Null extends Literal
+  constructor: -> super 'null'
```

Comment:
The ThisLiteral constructor is on two lines and this is on one line. Be consistent.